### PR TITLE
feat(server): add a template for queue service

### PR DESCRIPTION
**Pull request by tupe12334**
## PR Details

This pr is an open conversation about an Amplication standard for a queue service pr give you remarks and reviews

### CODE REVIEW

File: packages/amplication-server/src/core/queue/queue.module.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { Module } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { ClientsModule, Transport } from '@nestjs/microservices';
+import { QueueService, QUEUE_SERVICE_NAME } from './queue.service';
+
+export const BROKER_IP_ENV_KEY = 'BROKER_IP_ENV_KEY';
+
+@Module({
```
Comment (abrl91) [inline packages/amplication-server/src/core/queue/queue.module.ts:8]: why did you decide to go with '@Module' and not '@Client'?
https://docs.nestjs.com/microservices/kafka#client

**Comment by tupe12334** [comment]
Hi everyone, what do you think about a wrapper function that injects the emit data to value property, it let us add in the future more data:
```
{
value: the_data
// in the future
meta:{
sender:"amplication-server"
etc...
}
}
```
And it will let us a constant property with the request data
Would like to hear you opinion

File: packages/amplication-server/src/core/queue/queue.module.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            throw new Error('Missing broker ip in env file');
+          }
+          return {
+            transport: Transport.KAFKA,
+            options: {
+              client: {
+                clientId: 'server-queue-client',
+                brokers: [envBrokerIp]
+              }
+            }
```
Comment (abrl91) [inline packages/amplication-server/src/core/queue/queue.module.ts:25]: Don't we need to mention the consumer groupId?
`consumer: {
            groupId: 'consumer-group-id'
          }`

Comment (tupe12334) [reply packages/amplication-server/src/core/queue/queue.module.ts:25]: The groupId is only for consumers services

File: packages/amplication-server/src/core/queue/queue.module.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import { Module } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { ClientsModule, Transport } from '@nestjs/microservices';
+import { QueueService, QUEUE_SERVICE_NAME } from './queue.service';
+
+export const BROKER_IP_ENV_KEY = 'BROKER_IP_ENV_KEY';
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/queue/queue.module.ts:6]:
```suggestion
export const QUEUE_BROKER_URL_VAR = 'QUEUE_BROKER_URL';
```

File: packages/amplication-server/src/core/queue/queue.module.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        useFactory: (configService: ConfigService) => {
+          const envBrokerIp = configService.get<string>(BROKER_IP_ENV_KEY);
+          if (!envBrokerIp) {
+            throw new Error('Missing broker ip in env file');
+          }
+          return {
+            transport: Transport.KAFKA,
+            options: {
+              client: {
+                clientId: 'server-queue-client',
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/queue/queue.module.ts:22]: extract to const

File: packages/amplication-server/src/core/queue/queue.service.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import { Inject, Injectable } from '@nestjs/common';
+import { ClientKafka } from '@nestjs/microservices';
+
+export const QUEUE_SERVICE_NAME = 'QUEUE_SERVICE';
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/queue/queue.service.ts:4]: move to queue.module.ts

File: packages/amplication-server/src/core/queue/queue.service.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+@Injectable()
+export class QueueService {
+  constructor(
+    @Inject(QUEUE_SERVICE_NAME)
+    private readonly kafkaService: ClientKafka
+  ) {}
+  emitCreateGitPullRequest(data: any) {
+    return this.kafkaService.emit('git_pr_generate', data);
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/queue/queue.service.ts:13]: extract to const

File: packages/amplication-server/src/core/queue/queue.service.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+export const QUEUE_SERVICE_NAME = 'QUEUE_SERVICE';
+
+@Injectable()
+export class QueueService {
+  constructor(
+    @Inject(QUEUE_SERVICE_NAME)
+    private readonly kafkaService: ClientKafka
+  ) {}
+  emitCreateGitPullRequest(data: any) {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/queue/queue.service.ts:12]: use type

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-server/src/core/queue/queue.module.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import { Module } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { ClientsModule, Transport } from '@nestjs/microservices';
+import { QueueService, QUEUE_SERVICE_NAME } from './queue.service';
+
+export const QUEUE_BROKER_URL_VAR = 'QUEUE_BROKER_URL_VAR';
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/queue/queue.module.ts:6]:
```suggestion
export const QUEUE_BROKER_URL_VAR = 'QUEUE_BROKER_URL';
```

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-server/src/core/queue/queue.module.ts (hunk 4 above)
Comment (mylesgodino) [reply packages/amplication-server/src/core/queue/queue.module.ts]: Yes

**Pull request closed without merging**